R/mobr.R

```diff
@@ -10,221 +10,244 @@ require(pracma)
+    ind_S = rarefaction(sad, 'indiv', effort)
     return(ind_S)
 }
 
-get_deltaSsad = function(dat_sp, dat_plot, treatment1, treatment2) {
-    rescaled_ind = sapply(c(treatment1, treatment2), function(x) 
-                          rarefy_individual(dat_sp, dat_plot, x))
-    Nind = min(length(unlist(rescaled_ind[1])), length(unlist(rescaled_ind[2])))
-    deltaSsad = as.numeric(unlist(rescaled_ind[2])[1:Nind] - 
-                           unlist(rescaled_ind[1])[1:Nind])
+get_deltaSsad = function(dat_sp, dat_plot, groups, Nind=NULL) {
+    if (is.null(Nind))
+        Nind = min(tapply(rowSums(dat_sp),  dat_plot$group, sum))
+    # the first element of the argument groups must be the reference group
+    rescaled_ind = sapply(groups, function(x) 
+                          rarefy_individual(dat_sp, dat_plot, x, 1:Nind))
+    deltaSsad = rescaled_ind[ , 2] - rescaled_ind[ , 1]
     return(deltaSsad)
 }
 
-get_deltaSN = function(dat_sp, dat_plot, treatment1, treatment2, ScaleBy = NA) {
-    avg_dens = get_avg_dens(dat_sp, dat_plot, ScaleBy)
-    nplots = c(nrow(dat_plot[dat_plot[ , 2] == treatment1, ]), nrow(dat_plot[dat_plot[ , 
-        2] == treatment2, ]))
-    implicit_sample = sapply(c(treatment1, treatment2), function(x)
-                             rarefy_sample_implicit(dat_sp, dat_plot, x)[1:min(nplots)])
+get_deltaSN = function(dat_sp, dat_plot, groups, ScaleBy = NA, Nind=NULL) {
+    if (is.null(Nind))
+        Nind = min(tapply(rowSums(dat_sp),  dat_plot$group, sum))
+    avg_dens = get_avg_dens(dat_sp, dat_plot, ScaleBy) 
+    # shouldn't the above line be:  'mean(rowSums(dat_sp))' ?
+    effort = 1:min(table(dat_plot$group[dat_plot$group %in% groups]))
+    implicit_sample = sapply(groups, function(x)
+                             rarefy_sample_implicit(dat_sp, dat_plot, x, effort))
+    # DJM: It may be more conservative to aggregate the individual rarefaction result
+    # rather than interpolate the sample based differences to an artifical level of 
+    # precision 
+    # rescale sample based differences via intpolation to reflect numbers of individuals sampled
     # Use S_diff(1) = 0 for interpolation
-    implicit_sample_ind = pchip(c(1, (1:nrow(implicit_sample)) * avg_dens), 
-                                c(0, implicit_sample[ , 2] - implicit_sample[ , 1]),
-                                1:floor(nrow(implicit_sample) * avg_dens))
-    deltaSsad = get_deltaSsad(dat_sp, dat_plot, treatment1, treatment2)
-    # add variable called min index to clean up code
-    deltaSN = implicit_sample_ind[1:min(length(implicit_sample_ind), 
-                  length(deltaSsad))] - deltaSsad[1:min(length(implicit_sample_ind),
-                  length(deltaSsad))]
+    # here the differences are limited to comparisons of two groups
+    deltaSsamp = pchip(xi=c(1, effort * avg_dens), 
+                       yi=c(0, implicit_sample[ , 2] - implicit_sample[ , 1]),
+                       x=1:Nind)
```

Comment:
In the original code the interpolation goes to not necessarily to min(N_group) (which is what Nind is, I think), but to number of plots within group * avg_dens. But we can talk about which is more reasonable.